sphinx/domains/python.py

```diff
@@ -418,11 +434,11 @@ class PyObject(ObjectDescription[Tuple[str, str]]):
 
     allow_nesting = False
 
-    def get_signature_prefix(self, sig: str) -> str:
+    def get_signature_prefix(self, sig: str) -> List[nodes.Node]:
```

Comment:
Is this really part of the public API?

Reply:
No, it's a private one, I believe. But, as you know, we've seen many extensions and hacks that depend on private APIs. So this change might cause trouble. So it would be better not to change the interface as possible during the same major version (or let's declare it as an incompatible change).

Reply:
Right ([Hyrum's Law](https://www.hyrumslaw.com/) :-(). Even if this method is not changed, a new can't be introduced early either, as the new call in ``run()`` would be incompatible for anyone who subclassed ``PyObject``.
So, as far as I can see:
- Option 1: I can split the PR so some can be merged in 4.x, and the incompatible part in master.
- Option 2: Merge it all now to 4.x with an extra incompatibility note for anyone who messed with these private APIs.
- Option 3: Make a parallel version and hax ``run()`` to detect if the dynamic type of ``self`` is one of the normal Python objects and if not check whether the subclass overwrites the new method, and use that to call the right version. (This seems quite brittle and a lot of work to shield those that messed with private APIs.)

Reply:
Yes. My first proposal is option 1. But I admit it's a very cowardly approach. So I can agree with option 2 (Note: We have to mention it as incompatibility change in CHANGES). Both are okay to me.

Reply:
I have added an incompatibility note (option 2).

Reply:
Well, I wanted to add the note. _Now_ it should be pushed.